keyboard/src/lib.rs

```diff
@@ -1,11 +1,433 @@
 #![no_std]
 
-static CODES: [u8; 59] = *b"??1234567890-=??qwertyuiop[]\n?asdfghjkl;'`?\\zxcvbnm,./?*? ?";
+pub enum Action {
+    Pressed,
+    Released,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Key {
+    KA,
+    KB,
+    KC,
+    KD,
+    KE,
+    KF,
+    KG,
+    KH,
+    KI,
+    KJ,
+    KK,
+    KL,
+    KM,
+    KN,
+    KO,
+    KP,
+    KQ,
+    KR,
+    KS,
+    KT,
+    KU,
+    KV,
+    KW,
+    KX,
+    KY,
+    KZ,
+    K0,
+    K1,
+    K2,
+    K3,
+    K4,
+    K5,
+    K6,
+    K7,
+    K8,
+    K9,
+    KBacktick,
+    KDash,
+    KEqual,
+    KTab,
+    KBackslash,
+    KSpace,
+    KLeftSquareBracket,
+    KRightSquareBracket,
+    KSemicolon,
+    KApostrophe,
+    KComma,
+    KPeriod,
+    KSlash,
+    KKeypadAsterisk, // Keypad
+    KKeypadMinus,    // Keypad
+    KKeypadPlus,     // Keypad
+    KBackspace,
+    KEnter,
+    KLeftShift,
+    KRightShift,
+}
+
+impl Key {
+    pub fn is_ascii(self) -> bool {
+        self.ascii(false).is_some()
+    }
+
+    pub fn ascii(self, shift: bool) -> Option<char> {
+        use Key::*;
+
+        let character = match (self, shift) {
+                   (KA, false) => 'a',          (KA, true) => 'A',
```

Comment:
Was trying to make it a little easier to scan, but it's definitely non-standard. 😛  Reformatted in cc4a8fc to be more like the Rust standard.